shared-bindings/pixelbuf/PixelBuf.c

```diff
@@ -0,0 +1,506 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/objarray.h"
+#include "py/mphal.h"
+#include "py/runtime.h"
+#include "py/binary.h"
+#include "py/objproperty.h"
+
+#include <string.h>
+
+#include "PixelBuf.h"
+
+static pixelbuf_rgbw_t pixelbuf_byteorder_lookup[] = {
+    {0, 1, 2, 3},  // BYTEORDER_RGB
+    {0, 2, 1, 3},  // BYTEORDER_RBG
+    {1, 0, 2, 3},  // BYTEORDER_GRB
+    {1, 2, 0, 3},  // BYTEORDER_GBR
+    {2, 0, 1, 3},  // BYTEORDER_BRG
+    {2, 1, 0, 3},  // BYTEORDER_BGR
+};
+
+//|   .. attribute:: bpp
+//|
+//|     The number of bytes per pixel in the buffer (read-only)
+//|
+STATIC mp_obj_t pixelbuf_pixelbuf_obj_get_bpp(mp_obj_t self_in) {
+    pixelbuf_pixelbuf_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return mp_obj_new_int_from_uint(self->bpp);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(pixelbuf_pixelbuf_get_bpp_obj, pixelbuf_pixelbuf_obj_get_bpp);
+
+const mp_obj_property_t pixelbuf_pixelbuf_bpp_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&pixelbuf_pixelbuf_get_bpp_obj,
+              (mp_obj_t)&mp_const_none_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+
+//|   .. attribute:: brightness
+//|
+//|     Float value between 0 and 1.  Output brightness.
+//|     If the PixelBuf was allocated with two both a buf and a rawbuf,
+//|     setting this value causes a recomputation of the values in buf.
+//|     If only a buf was provided, then the brightness only applies to
+//|     future pixel changes.
+//|
+STATIC mp_obj_t pixelbuf_pixelbuf_obj_get_brightness(mp_obj_t self_in) {
+    pixelbuf_pixelbuf_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return mp_obj_new_float(self->brightness);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(pixelbuf_pixelbuf_get_brightness_obj, pixelbuf_pixelbuf_obj_get_brightness);
+
+
+STATIC mp_obj_t pixelbuf_pixelbuf_obj_set_brightness(mp_obj_t self_in, mp_obj_t value) {
+    pixelbuf_pixelbuf_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    self->brightness = mp_obj_float_get(value);
+    if (self->brightness > 1)
+        self->brightness = 1;
+    else if (self->brightness < 0)
+        self->brightness = 0;
+    if (self->two_buffers)
+        pixelbuf_recalculate_brightness(self);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(pixelbuf_pixelbuf_set_brightness_obj, pixelbuf_pixelbuf_obj_set_brightness);
+
+const mp_obj_property_t pixelbuf_pixelbuf_brightness_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&pixelbuf_pixelbuf_get_brightness_obj,
+              (mp_obj_t)&pixelbuf_pixelbuf_set_brightness_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+void pixelbuf_recalculate_brightness(pixelbuf_pixelbuf_obj_t *self) {
+    for (uint i = 0; i < self->bytes; i++) {
+       self->buf[i] = self->rawbuf[i] * self->brightness;
+    }
+}
+
+//|   .. attribute:: buf
+//|
+//|     The bytearray of pixel data after brightness adjustment
```

Comment:
Why is this needed externally?

Reply:
rawbuf is the one that's not needed externally.  buf is the bytearray needed for sending to neopixels.